### DISCUSSION TITLE
OpenDCT: An open source digital cable tuner.

### DISCUSSION THREAD

**Discussion by Ugadata** [OpenSageTV/Community#30] [OpenDCT]
I copied and pasted EnterNoEscape's 1st post about OpenDCT from the original forums.  There are certainly other post in this thread (and others) that would be very helpful (IMO) going forward to be here.

I hope it is helpful and I apologize if this should be done a different way.
Maybe just static links pointing to the various posts?

> OpenDCT: An open source digital cable tuner.
> OpenDCT is now available to the public! 
> 
> The latest stable installer can be found on GitHub.
> 
> The project source code and some useful installation and configuration information can be found on GitHub.
> 
> The minimum Java version is 1.7.
> 
> InfiniTV 4 users must be on the very last beta firmware from Ceton to be able to use OpenDCT.
> 
> On Windows, I recommend the 32-bit (x86) version unless you actually have 64-bit Java runtime installed. If you don't have 64-bit runtime installed and you install the 64-bit version of OpenDCT, you will not be able to start OpenDCT successfully. If you run into issues with Java Service Wrapper being unable to find your Java runtime, you will need to edit wrapper.conf under C:\Program Files [(x86)]\OpenDCT Authors\OpenDCT\jsw\conf\wrapper.conf to contain the complete path to your installed runtime. Most people should not need to do this.
> 
> On Linux, I recommend downloading the architecture of your distribution.
> 
> For the first run on any platform run OpenDCT on the console first. This is because it needs to detect your devices and populate the opendct.properties file. It will also immediately show if you have problems. On Linux there is a script called ./console-only and the configuration files can be found under /etc/opendct/conf. On versions 0.4.17-Beta and earlier configuration files are located under /opt/opendct/conf. On Windows, there is a shortcut called OpenDCT Run as Console and the configuration can be accessed by the shortcut OpenDCT Properties. Press Ctrl-C to stop the console after you see that it found your devices or just wait 30 seconds, then press Ctrl-C. It will shutdown cleanly. The next major release will replace manually editing the properties file with a web interface. After you have done this, take a look at the Configuration section on GitHub for details on how you can limit the devices that OpenDCT will initialize and how to configure SageTV to discover OpenDCT.
> 
> If you're having trouble getting started, many people have found this post by Daweeze very helpful.
> 
> The Windows installer will do upgrades so it is unnecessary to uninstall before installing a new version. It is however necessary to stop the service before upgrading. This post by KeithAbbott should help you out if you're interested in backing things up before upgrading. If you delete opendct.properties, make sure you make a note of what you have changed before you delete it.
> 
> Here are a few properties not on the GitHub page that might look interesting with their default values included if possible. This is not an exhaustive list, but it covers the basics.
> sagetv.device.global.required_devices_loaded_timeout_ms=30000
> This is the amount of time allowed in milliseconds to pass while waiting for the required number of devices before the program will exit with a failure. This timeout will only happen if sagetv.device.global.required_devices_loaded_count is higher than 0.
> sagetv.device.global.required_devices_loaded_count=0
> This is the number of devices that need to be detected and loaded before the required timeout. This must be the total number of devices OpenDCT normally detects and loads for standby support to work correctly. If this value is left at zero, OpenDCT will just return failures to SageTV when it asks for a device that has not been detected yet because it has no way to know if it should have that device or not. If you are not using standby changing this parameter will also help at startup because it will make OpenDCT hang onto any tuning requests from SageTV until all of the devices have been loaded.
> hdhr.enabled=false
> If set true, this will enable HDHomeRun native discovery and at the moment doesn't do anything because we detect all devices currently via UPnP even if that is not the method we will actually use for tuning.
> hdhr.locking=true
> If set true, this will enable HDHomeRun locking. This will only work if the HDHomeRun device is being tuned using its native protocol. The native protocol is used by default.
> sagetv.device.<unique_id>.always_force_external_unlock=false
> If set true, and locking is enabled, this will force the requested HDHomeRun device to become locked by OpenDCT. This will only work if the HDHomeRun device is being tuned using its native protocol. The native protocol is used by default.
> pm.enabled=true
> If set true, only on Windows, this will allow OpenDCT to gracefully handle the computer entering and exiting standby. There is provisioning for other operating systems, but no actual implementation.
> sagetv.device.<unique_id>.device_name=
> You can change the names of any discovered device. You must be careful to not rename a device to the same name as another device. This will result in the first discovered device loading and the second discovered device not loading.
> sagetv.device.parent.<unique_id>.consumer=opendct.consumer.FFmpegTransSageTVConsumerImpl
> This is the implementation to be used for streaming to SageTV. The currently valid values for this property are the default, opendct.consumer.RawSageTVConsumerImpl and opendct.consumer.MediaServerConsumerImpl (new in 0.5 and only uses media server). Parent entries cover an entire device. Right now this is for convenience, so you don't need to set this value for each tuner. It will probably change once we are using a web interface. opendct.consumer.FFmpegSageTVConsumerImpl has been deprecated as of 0.4.
> sagetv.device.parent.<unique_id>.http.producer=opendct.producer.HTTPProducerImpl (0.5 default: opendct.producer.NIOHTTPProducerImpl)
> This is the implementation to be used for receiving content from devices with HTTP capabilities. opendct.producer.NIOHTTPProducerImpl was added as of 0.5.
> sagetv.device.parent.<unique_id>.lineup=
> These lineups are automatically generated based on the available devices. This allows you to do things like use the lineup from you InfiniTV device that actually has a CableCARD to look up the frequency and program for any ClearQAM device. The lineup files are located under lineups in the configuration directory and are updated by default every 8 hours.
> sagetv.device.parent.<unique_id>.local_ip_override=
> This is the local interface IP address. It is not populated by default. If OpenDCT is detecting your local IP address that communicates with this remote device incorrectly, you can override it here.
> consumer.ffmpeg.upload_id_enabled=false
> consumer.raw.upload_id_enabled=false
> If set true, these will use the Media Server (upload id) feature of SageTV instead of writing directly to the file when recording.
> sagetv.use_automatic_loopback=true
> If set true, OpenDCT will detect if it is running on the same computer as SageTV and use the loopback interface when sending PROPERTIES.
> upnp.dct.fast_tuning=false (no longer available in 0.5)
> If set true, this will enable the tuner to always be in PLAY mode with RTSP configured. When the command STOP is received, the channel is tuned to 0 or if the device is ClearQAM, it will tune to the frequency 0. This improves tuning time by as much as 500ms.
> upnp.dct.http_tuning=true (no longer available in 0.5)
> If set true, this uses the same "tricks" as fast_tuning, but this uses the forms on the web interface to tune into channels instead of the UPnP interface. This method is even faster. This method is also the only way to tune when using an InfiniTV device in ClearQAM mode (no CableCARD). Even if the value is false, it will be overridden just for ClearQAM. This is only used on InfiniTV devices. It has no effect on tuning the HDHomeRun Prime.
> upnp.dct.hdhr_tuning=true (no longer available in 0.5)
> If set true, this will use the HDHomeRun native protocols for tuning. This is the fastest possible way to tune the HDHomeRun Prime. This is the only method used to tune the Prime when it is in ClearQAM mode. Even if the value is false, it will be overridden just for ClearQAM. This setting has no effect on InfiniTV devices.
> upnp.qam.automap_reference_lookup=true (replaced the the global option channels.qam.automap_reference_lookup=true in 0.5)
> If set true, ClearQAM devices will check all available lineups for frequency and program information if they don't already have it and then update their own lineup with the discovered information when a channel is requested.
> upnp.qam.automap_tuning_lookup=false (replaced the the global option channels.qam.automap_tuning_lookup=true in 0.5)
> If set true, ClearQAM devices after not being able to find the frequency and program for the requested channel on any lineup, it will select a currently not in use tuner that also has a CableCARD installed to tune the channel and get the frequency and program that way. If all tuners are in use, it will fail.
> 
> New Features in 0.4
> Tuner Pooling.
> This new feature enables OpenDCT to intelligently select a tuner based on merit and lock status in that order. It does two passes over the tuners. The first one tries to find the highest merit tuner that is not currently in use by SageTV; it checks each one in order to see if it's locked. If they are all either in use or locked, it then it goes through a shorter list of tuners not in use by SageTV, but locked and checks if it's allowed to unlock any of them forcefully. If it's not allowed on any of them, then it gives up. These locked checks are surprisingly fast. Since the InfiniTV devices do not have a locked indication like the HDHomeRun Prime devices, enabling pooling with only InfiniTV devices will just remap them based on merit.
> To enable the pooling feature change the property pool.enabled from false to true.
> Because of this new feature, the force unlock feature is now set per capture device. To enable forced unlocking, change the value of sagetv.device.<unique_id>.always_force_external_unlock to true for each device that you want this feature enabled on. The global setting no longer has any effect.
> Merit is set by changing the value of sagetv.device.<unique_id>.encoder_merit. Higher numbers mean higher priority. This is also the merit that will be sent on discovery.
> Pools are automatically assigned based on capture device type the first time you run OpenDCT. You can change the pools by changing the value of sagetv.device.<unique_id>.encoder_pool to another name for all of the capture devices that you want to put into a different pool. The name is case sensitive. Also be sure to not have any extra spaces. I highly do not recommend putting capture devices that cannot tune the same channels into the same pool. It will create a problem per paragraph two in the issue created for this feature.
> Early Port Assignment.
> When this is enabled, OpenDCT will immediately (< 500ms from startup) open any ports assigned in opendct.properties. For most people this will just be port 9000. Set the property sagetv.early_port_assignment to true to enable this feature.
> If you are using this feature, just like standby, you must set sagetv.device.global.required_devices_loaded_count to match the total number of capture devices OpenDCT should expect to be detected and loaded. If you do not set this property and SageTV makes a tuning request before anything has been detected, OpenDCT will not know that it should hang onto that request and the request will just return ERROR to SageTV.
> HDHomeRun Prime Channel Auto-Enabling.
> This is enabled by default. It can be disabled by setting channels.prime.enable_all_channels to false.
> The process of importing these channels already happens one minute after startup and every 8 hours by default in OpenDCT. The change is that now unless the channel actually indicates that it required DRM to tune, OpenDCT assumes it's tunable. What this translates into directly is you can now do a channel scan within SageTV and import your last channel scan from your HDHomeRun Prime.
> UPnP Ignore Interface by IP or Interface Name
> You can now specify a comma separated list of local IP addresses or interfaces to prevent UPnP from using them.
> You can provide a comma separated list of interfaces on the property upnp.service.configuration.ignore_interfaces_csv. You will need to reference the log to figure out what interfaces you want to list here. The ones being used are listed very early in opendct.log. This difficulty is the reason why the IP address option is available too.
> You can provide a comma separated list of IP addresses on the property upnp.service.configuration.ignore_local_ip_csv.
> Exclusive Server Address.
> If you set the property sagetv.device.<unique_id>.exclusive_server_address to the IP address of the server intended to use it, when PROPERTIES are requested if the request did not come from this IP address, the capture device is not returned. This should help some people clean up tuner discovery, since you can now effectively make specific tuners only visible to a specific server.
> Transcoding.
> Set the software transcoding profile by changing sagetv.device.<unique_id>.transcode_profile to one of the available profiles are stored under C:\ProgramData\OpenDCT\config\transcode on Windows and /etc/opendct/conf/transcode on Linux. See profile_example.properties for help on how you can create your own. Do not include the .properties extension when setting the profile.
> 
> New Features in 0.5
> 
> CCExtractor Options:
> consumer.ffmpeg.ccextractor_all_streams=true
> This enables extracting of both CC1 and CC2. If this is off, it will output only CC1. You will likely want to leave this set to true.
> consumer.ffmpeg.ccextractor_custom_options=
> This can be any combination of valid CCExtractor options with the exception of -12, -1, -2, -in=, -out= and -o. Some people might want to change how roll-up captions are written for example. These options are sanitized to prevent execution of other programs, so you will not be able to do that from this option.
> consumer.ffmpeg.ccextractor_enabled=false
> This enables CCExtractor when set to true.
> 
> Dynamic Consumer Options:
> Channel Range Notation
> Channel ranges are defined by a number then a hyphen followed by another number (8-10). Multiple ranges are separated by a comma (13-15,8-10). ATSC channels are defined by a number (frequency channel) then a hyphen followed by a the channel number then a hyphen followed by the sub-channel number (50-8-1). The use of two hyphens automatically makes it literally a channel instead of a range.
> Ex. 8-10,50-8-1,13-16 will match the following channels: 8, 9, 10, 50-8-1, 13, 14, 15, 16
> consumer.dynamic.channels.ffmpeg=
> These are the channels that will always be assigned to the transcoding FFmpeg consumer.
> consumer.dynamic.channels.raw=
> These are the channels that will always be assigned to the raw consumer.
> consumer.dynamic.default=opendct.consumer.FFmpegTransSageTVConsumerImpl
> This is the consumer always used if the channel currently tuned doesn't match the channels under consumer.dynamic.channels.ffmpeg or consumer.dynamic.channels.raw.
> 
> Generic HTTP Capture Device:
> Creating Capture Devices for Editing
> Create names for them under the property generic.http.device_names_csv in opendct.properties separated by commas. Start the OpenDCT service, let it run for a few seconds, then stop it. You should now see entries in opendct.properties that have the property sagetv.device.<unique_id>.device_name matching the name of the devices you requested. The names can really be anything, just don't use commas in the names for what should be obvious reasons.
> Ex. generic.http.device_names_csv=Encoder 1,Encoder 2
> sagetv.device.<unique_id>.channel_padding=
> This is the minimum length to be passed for the %c% variable. Values shorter than this length will have zeros (0) appended to the left of the channel to make up the difference. (Ex. 8 becomes 008 if this is set to 3.)
> sagetv.device.<unique_id>.custom_channels=
> This is an optional semicolon delimited list of channels you want to appear in SageTV for this device. This is a shortcut around creating an actual OpenDCT lineup. If there are any values in the field, they will override the lineup assigned to this capture device on a channel scan. This provides an easy way to add channels if you are not actually going to use guide data.
> sagetv.device.<unique_id>.pretuning_executable=
> This is an optional executable that if defined, will always be run before actually tuning the channel. You can add the channel as an argument by using the variable %c%. Don't forget to escape backslashes (\ needs to be \\).
> sagetv.device.<unique_id>.stopping_executable=
> This is an optional executable that if defined, will always be run when the capture device is told to stop. You can add the last tuned channel as an argument by using the variable %c%. Don't forget to escape backslashes (\ needs to be \\).
> sagetv.device.<unique_id>.streaming_url=
> This is a URL that points directly to an audio/video stream. If you provide %c% in the URL, it will be replaced with the channel requested by SageTV. HLS and m3u8 playlists are not supported at this time.
> sagetv.device.<unique_id>.streaming_url2=
> This is an alternative URL that points directly to an audio/video stream. If you provide %c% in the URL, it will be replaced with the channel requested by SageTV. This stream will only be used if a channel matches one of the ranges in streaming_url2_channels. HLS and m3u8 playlists are not supported at this time.
> sagetv.device.<unique_id>.streaming_url2_channels=
> This is the channel ranges that will use the alternative URL. This uses the same formatting as the dynamic consumer channel ranges.
> sagetv.device.<unique_id>.tuning_delay_ms=0
> This is the amount of time in milliseconds to wait after the program associated with tuning_executable has returned.
> sagetv.device.<unique_id>.tuning_executable=
> This is an optional executable that if defined will be used to change the channel being streamed. Insert %c% where the channel needs to be provided to the executable. If %c% isn't provided, but this property is defined, the channel number will be appended as a final parameter. Don't forget to escape backslashes (\ needs to be \\).
> Ex. sagetv.device.<unique_id>.tuning_executable=/full/path/tune 0 %c%
> Ex. sagetv.device.<unique_id>.tuning_executable=C:\\Full\\Path\\tune.exe 0 %c%
> Ex. sagetv.device.<unique_id>.tuning_executable=cmd /c C:\\Full\\Path\\tune.cmd 0 %c%

#

**Comment by JustFred51** [comment]
And here's the super-helpful posting by Daweeze, with his concise setup info:

[https://forums.sagetv.com/forums/showthread.php?p=582155&postcount=175](url)
> Ok, so here's the latest after I uninstalled OpenDCT 0.3.13-beta (Also manually removed folder under C:\ProgramData).
> 
> For Reference:
> 
> Operating System: Windows 8.1
> SageTV Version: 7.1.9.256
> Java Version: 1.8.0_66 (32-bit)
> Network Tuners: 2x HDHomeRun Prime (CableCard) = 6 Tuners
> 
> Installation:
> 
> 1) Downloaded OpenDCT 0.3.14-beta.
> 2) Stopped SageDCT Service, Changed Startup Type to Disabled.
> 3) Stopped SageTV Service.
> 4) Made Backup Copy of Sage.properties and Wiz.bin.
> 5) Installed OpenDCT 0.3.14-beta.
> 6) Ran "OpenDCT Run as Console".
> 7) CTRL-C on OpenDCT Following Download of Channel Lineups.
> 8) Ran "OpenDCT Start Service" - Success!
> 9) Edited and Saved Sage.properties and changed network_encoder_discovery=false to network_encoder_discovery=true.
> 10) Started SageTV Service.
> 11) Started SageTV.
> 12) Setup>Setup Video Sources>Add New Source.
> 13) Selected New Video Capture Device (Ex. DCT-HDHomeRun Prime Tuner 13113C69-0 on 127.0.0.1:9000).
> 14) Selected "Digital TV Tuner" (Box Turns Red).
> 15) Selected "Digital TV Tuner" Again (Message: "The Broadcast Standard is being determined. Please wait...").
> 16) Selected "Use Tuner on Capture Card to tune Antenna/Cable TV. (To select between Antenna of Cable, use the SageTV installation that is controlling this Network Encoder)".
> 17) Selected "Use or Copy an Existing Channel Lineup for this Source".
> 18) Selected the Existing Lineup (Ex. Comcast Digital - Fairfield).
> 19) Selected "Use the same channel lineup".
> 20) Repeated Steps 12-19 For Each Remaining OpenDCT Tuner Available.
> 21) Exited SageTV.
> 22) Stopped SageTV Service.
> 23) Edited and Saved Sage.properties and Removed the "mmc/encoders" Entries For Each SageDCT Tuner Identified. (Ex. "mmc/encoders/123456990/video_capture_device_name=SageDCT-HDHomeRun Prime Tuner 13113C69-1" is the Identifying Entry That Ties My SageDCT Tuner to #1234560990, So All "mmc/encoders/1234560990....." Entries Were Deleted As Were 5 Other SageDCT Tuner Entries.
> 24) Started SageTV Service.
> 25) Started SageTV.
> 26) Setup>Setup Video Sources>Add New Source.
> 27) Confirmed only OpenDCT Tuner Entries Were Present.
> 28) Tested Recording On All 6 Tuners Simultaneously - Success!
> 
> Notes:
> 
> 1) Default configuration worked fine, no modifications of any file were necessary.
> 2) OpenDCT Tuners show "DCT-TunerName Etc....." as a Source. I expected "OpenDCT-TunerName Etc....." as a Source. I see no reason why to not do that since SageDCT does it as well.
> 
> 
> So far so good, thanks for the hard work by everyone on this project and helping keep SageTV alive. I have high hopes with this release!!
> 
> Also, I have attached a copy of the Channel Lineup that was bombing out in the 0.3.13-beta in case it helps clarify what a Lineup might return and why.
> 
> Thanks again!
>

**Comment by jusjoken** [comment]
I have pinned this information to the OpenDCT category.  Please open new threads for further discussions

**Comment by Robert-Rathbone** [comment]
I'm having trouble getting an InfiniTV 6 PCIe tuner working with OpenDCT, so hopefully someone with experience with that tuner card can help me.  I asked for help in another forum category (https://github.com/OpenSageTV/Community/discussions/33) but I hope that here I can get more attention from people who can help.

I reinstalled the OS with an earlier version (Ubuntu Server 20.04.6) in hopes that it would start working for me there since that was the version I previously had it running under (I think).  I was able to get the source code to compile for the driver (kernel module) found at (https://github.com/JamesRHarris/infinitv_pcie) with some adjustments to the code.

The problem is the same as I had before: OpenDCT is not detecting the tuner via uPnP.  It sees the 192.168.200.2 address I gave it but the tuner is apparently not responding with any uPnP discovery requests.  The tuner responds to pings and uPnP requests on 192.168.200.1 but OpenDCT doesn't recognize that IP address nor does it appear when I run "ifconfig -a".  I don't know how to get the 192.168.200.1 address to show up so that OpenDCT can use it.  When I've tried assigning the ctn0 interface to that IP address, it replaces the 192.168.200.1 with the channel that was running on 192.168.200.2.  

I'm at the end of my rope.  Does anyone have any ideas?

Thanks,
Robert
